### ISSUE TITLE
Implement first-class functions

### ISSUE THREAD

**Issue by jfecher** [noir-lang/noir#467]
# Problem

Noir currently has generics, but no way to abstract over operations to perform on these generics. With the absence of traits, another method to provide functionality is by manually passing in functions. This requires higher-order functions in noir, or in other words, for functions to accept other functions as parameters and return them as return values.

# Solution

Add higher-order functions including function types and lambdas to Noir.

Function types:
```rust
fn()
fn(Field) -> Field
fn(i32, T) -> (T, i32)

// function takes a function and returns a function
fn(fn(T) -> U) -> fn(T) -> U
```

Lambdas:
```rust
|| {}
|x| x
|x, t| (x, t)
|f: fn(T) -> U| f
```

**Comment by kevaundray** [comment]
@jfecher can this be closed?

**Comment by jfecher** [comment]
Yes, implemented in #468
